src/charm.py

```diff
@@ -0,0 +1,245 @@
+    INIT_PATH,
+)
+from log import log_event_handler
+from state import State
+from utils import generate_password, push_files
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+
+class SupersetK8SCharm(CharmBase):
+    """Charm the service.
+
+    Attrs:
+        _state: used to store data that is persisted across invocations.
+        external_hostname: DNS listing used for external connections.
+    """
+
+    @property
+    def external_hostname(self):
+        """Return the DNS listing used for external connections."""
+        return self.config["external-hostname"] or self.app.name
+
+    def __init__(self, *args):
+        """Construct.
+
+        Args:
+            args: Ignore.
+        """
+        super().__init__(*args)
+        self.name = APP_NAME
+        self._state = State(self.app, lambda: self.model.get_relation("peer"))
+
+        # Handle basic charm lifecycle
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(
+            self.on.superset_pebble_ready, self._on_pebble_ready
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.restart_action, self._on_restart)
+
+        # Handle Ingress
+        self._require_nginx_route()
+
+    def _require_nginx_route(self):
+        """Require nginx-route relation based on current configuration."""
+        require_nginx_route(
+            charm=self,
+            service_hostname=self.external_hostname,
+            service_name=self.app.name,
+            service_port=APPLICATION_PORT,
+            tls_secret_name=self.config["tls-secret-name"],
+            backend_protocol="HTTPS",
+        )
+
+    @log_event_handler(logger)
+    def _on_install(self, event):
+        """Install application.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = MaintenanceStatus(f"installing {APP_NAME}")
+
+    @log_event_handler(logger)
+    def _on_pebble_ready(self, event: PebbleReadyEvent):
+        """Define and start a workload using the Pebble API.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self._update(event)
+
+    @log_event_handler(logger)
+    def _on_config_changed(self, event: ConfigChangedEvent):
+        """Handle changed configuration.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = WaitingStatus(f"configuring {APP_NAME}")
+        self._update(event)
+
+    def _restart_application(self, container):
+        """Restart application.
+
+        Args:
+            container: application container
+        """
+        self.unit.status = MaintenanceStatus(f"restarting {APP_NAME}")
+        container.restart(self.name)
+        self.unit.status = ActiveStatus()
+
+    def ready_to_start(self):
```

Comment:
Is this method necessary since it is only checking for ```self._state.is_ready()```? Seems like you can just check directly without introducing this method.